scipy/spatial/_hausdorff.pyx

```diff
@@ -0,0 +1,63 @@
+"""
+Directed Hausdorff Code
+
+.. versionadded:: 0.19.0
+
+"""
+#
+# Copyright (C)  Tyler Reddy, Richard Gowers, and Max Linke, 2016
+#
+# Distributed under the same BSD license as Scipy.
+#
+
+import numpy as np
+cimport numpy as np
+cimport cython
+from libc.math cimport sqrt
+
+__all__ = ['directed_hausdorff']
+
+@cython.boundscheck(False)
+def directed_hausdorff(double[:,::1] ar1, double[:,::1] ar2):
+
+    cdef double cmax, cmin, d
+    cdef int break_occurred
+    cdef int N1 = ar1.shape[0]
+    cdef int N2 = ar2.shape[0]
+    cdef int data_dims = ar1.shape[1]
+    cdef unsigned int i, j, k
+    cdef unsigned int i_store, j_store, i_ret, j_ret = 0
+    cdef long[:] resort1, resort2
+
+    # shuffling the points in each array generally increases the likelihood of
+    # an advantageous break in the inner search loop and never decreases the
+    # performance of the algorithm
+    resort1 = np.arange(N1)
+    resort2 = np.arange(N2)
+    np.random.shuffle(resort1)
+    np.random.shuffle(resort2)
```

Comment:
@tylerjereddy the problem is that `np.random.<function>` uses a **global singelton** instance of `np.random.RandomState`.  Everytime this distance function is now called the you change that global state. Consider the two loops below.

```python
# first version no Hausdorff distance calcuation
np.random.seed(123456)
for i in range(N):
    xyz = np.random.uniform(size=(100, 3))
    print(xyz.mean())

# with hausdorrf distance
np.random.seed(123456)
for i in range(N):
    xyz = np.random.uniform(size=(100, 3))
    directed_hausdorff(xyz[:50], xyz[50:])
    print(xyz.mean())
```

As a user I would assume that both loops print the same values. But since you change the global RandomState in `directed_hausdorff` they will be different and I can't reproduce my results anymore because I called a distance function.

Therefore using a local RandomState as @ev-br suggests will be better. It will also allow to call this in parallel using joblib. Modifying the np.random state in a function called with `joblib` can lead to more correlated random numbers. If you are lucky you will see right away that your results are off.